Modules/Intent.Modules.AspNetCore.MultiTenancy/FactoryExtensions/AspNetCoreIntegrationExtension.cs

```diff
@@ -17,303 +19,386 @@
-                    return;
-                }
-                
-                template.AddNugetDependency(new NugetPackageInfo("Finbuckle.MultiTenant", "6.5.1"));
-                template.AddNugetDependency(new NugetPackageInfo("Finbuckle.MultiTenant.EntityFrameworkCore", "6.5.1"));
+        return true;
+    }
 
-                template.CSharpFile.AfterBuild(file =>
-                {
-                    var (castTo, indexer) = hasMultipleConnectionStrings
-                        ? ($"({template.GetTypeName(TenantExtendedInfoTemplate.TemplateId)})", $"[\"{connectionStringNameInternal}\"]")
-                        : (string.Empty, string.Empty);
+    private static void GetSharedDatabaseDataIsolationConfiguration(
+        bool hasMultipleConnectionStrings, 
+        IApplication application, 
+        string connectionStringNameInternal, 
+        ICSharpFileBuilderTemplate dbContextTemplate)
+    {
+        if (!application.Settings.GetMultitenancySettings().DataIsolation().IsSharedDatabase())
+        {
+            return;
+        }
 
-                    var method = file.Classes.First().FindMethod("AddInfrastructure");
-                    if (method == null)
-                    {
-                        return;
-                    }
+        dbContextTemplate.CSharpFile.AfterBuild(file =>
+        {
+            file.AddUsing("Finbuckle.MultiTenant");
+            file.AddUsing("Finbuckle.MultiTenant.EntityFrameworkCore");
+            
+            var priClass = file.Classes.First();
+            priClass.ImplementsInterface("IMultiTenantDbContext");
+            
+            var ctor = priClass.Constructors.First();
+            ctor.AddParameter("ITenantInfo", "tenantInfo");
+            ctor.AddStatement("TenantInfo = tenantInfo;");
+
+            priClass.AddProperty("ITenantInfo", "TenantInfo", prop => prop.PrivateSetter());
+            priClass.AddProperty("TenantMismatchMode", "TenantMismatchMode", prop => prop.WithInitialValue("TenantMismatchMode.Throw"));
+            priClass.AddProperty("TenantNotSetMode", "TenantNotSetMode", prop => prop.WithInitialValue("TenantNotSetMode.Throw"));
+
+            var syncSave = priClass.FindMethod(p => "SaveChanges".Equals(p.Name) && p.Parameters.Count == 1);
+            if (syncSave == null)
+            {
+                priClass.AddMethod("int", "SaveChanges", method => method
+                    .Override()
+                    .AddParameter("bool", "acceptAllChangesOnSuccess")
+                    .AddStatement("return base.SaveChanges(acceptAllChangesOnSuccess);"));
+                syncSave = priClass.FindMethod(p => "SaveChanges".Equals(p.Name) && p.Parameters.Count == 1);
+            }
+            
+            var asyncSave = priClass.FindMethod(p => "SaveChangesAsync".Equals(p.Name) && p.Parameters.Count == 2);
+            if (asyncSave == null)
+            {
+                priClass.AddMethod("Task<int>", "SaveChangesAsync", method => method
```

Comment:
How does this interact with EF's `Task<int> SaveChangesAsync(CancellationToken cancellationToken)` which is overridden by our domain events pattern?

Reply:
![image](https://user-images.githubusercontent.com/7236289/223038668-8af59208-a090-47f8-8773-bc773c350694.png)
![image](https://user-images.githubusercontent.com/7236289/223038681-cbcb9f60-b685-4511-828d-62b3ae35f8a3.png)
![image](https://user-images.githubusercontent.com/7236289/223038710-2c980bb1-249f-4bc8-b9ed-c0620fb86bcd.png)
This is where it currently stands when DomainEvents get installed alongside Finbuckle. In theory this should still work since the SaveChangesAsync(CancellationToken) will invoke the other version that has "acceptAllChangesOnSuccess" in the end.

Reply:
I've confirmed that it gets called as expected so unless we want to modify the DomainEvents interop module, this as-is will work.